files/image_config/hostcfgd/hostcfgd

```diff
@@ -0,0 +1,339 @@
+
+        # Set local and tacacs+ pam priority
+        # ['local'] or ['local','tacacs+'] or ['tacacs+', 'local'] or ['tacacs+']
+        if self.auth_login == [AAA_LOCAL]:
+            pam_modules = [AAA_LOCAL]
+        elif self.auth_login == [AAA_LOCAL, AAA_TACPLUS]:
+            pam_modules = [AAA_LOCAL] + self.tacplus_servers
+        elif self.auth_login == [AAA_TACPLUS] or self.auth_login == [AAA_TACPLUS, AAA_LOCAL]:
+            # Don't accept only TACACS+ authentication
+            # Make sure root will always authentication on local, not TACACS+
+            pam_modules = self.tacplus_servers + [AAA_LOCAL]
+            auth_file_body += "auth\t[success=%d new_authtok_reqd=done default=ignore]\t" % (len(pam_modules)-1)
+            auth_file_body += "pam_succeed_if.so user = root debug\n"
+
+        if len(pam_modules):
+            for pam_module in pam_modules[:-1]:
+                line = self.get_pam_auth(pam_module)
+                if line is not "":
+                    auth_file_body += "auth\t" + pam_control + "\t" + line
+
+            pam_module = pam_modules[-1]
+            line = self.get_pam_auth(pam_module)
+            auth_file_body += "auth\t[success=1 default=ignore]\t" + line
+
+        with open(PAM_AUTH_FILE, "w") as f:
+            f.write(AUTH_FILE_HEADER + auth_file_body + AUTH_FILE_FOOTER)
+
+        # Modify common-auth include file in /etc/pam.d/type and sshd
+        if os.path.isfile(PAM_AUTH_FILE):
+            os.system("sed -i -e '/^@include/s/common-auth$/common-auth-sonic/' /etc/pam.d/sshd")
+            os.system("sed -i -e '/^@include/s/common-auth$/common-auth-sonic/' /etc/pam.d/login")
+        else:
+            os.system("sed -i -e '/^@include/s/common-auth-sonic$/common-auth/' /etc/pam.d/sshd")
+            os.system("sed -i -e '/^@include/s/common-auth-sonic$/common-auth/' /etc/pam.d/login")
+
+    # Set tacacs+ server in nss-tacplus conf
+    def modify_nss_conf(self):
+        if AAA_TACPLUS in self.auth_login:
+            # Add tacplus in nsswitch.conf if TACACS+ enable
+            if os.path.isfile(NSS_CONF):
+                os.system("sed -i -e '/tacplus/b' -e '/^passwd/s/compat/& tacplus/' /etc/nsswitch.conf")
+        else:
+            if os.path.isfile(NSS_CONF):
+                os.system("sed -i -e '/^passwd/s/ tacplus//' /etc/nsswitch.conf")
+
+        dbg = ""
+        servers = ""
+        if self.debug:
+            dbg = "debug=on\n"
+        for server in self.tacplus_servers:
+            servers += server.get_nss_conf(self.tacplus_global)
+
+        contents = ""
+        with open(NSS_TACPLUS_CONF, 'r') as f:
+            line = f.readline()
+            while line:
+                if '#' is line[0]:
+                    contents += line
+                elif 'debug=' != line[0:6] and 'server=' != line[0:7]:
+                    contents += line
+                line = f.readline()
+        contents += dbg + servers
+        with open(NSS_TACPLUS_CONF, 'w') as f:
```

Comment:
once you change the conf file, how to make the conf effective? do you need to reload anything?

Reply:
If tacplus is enabled for passwd in nsswitch.conf, the conf is loaded by nss-tacplus plugin when the function getpwnam() is called each time. Don't need to reload it.